tests/app/test_mqtt_reader.py

```diff
@@ -70,20 +73,26 @@ def test_fail_connect(self):
             self.mqtt_reader.run()
         assert "Cannot connect to MQTT broker" in str(ex.value)
 
-    def test_fail_connect_rc(self):
-        def call_on_connect(*args):
-            callback = self.mqtt_reader._on_connect()
-            callback(self.mock_mqtt_client, None, None, 1)
-            self.mock_error_handler.publish.assert_called()
-
-        self.mock_mqtt_client.connect.side_effect = call_on_connect
-        self.mqtt_reader.run()
-        self.mqtt_reader.stop()
-        self.mock_mqtt_client.connect.side_effect = None
-
     def test_disconnect(self, caplog):
         self.mock_mqtt_client.connect.return_value = 0
         self.mqtt_reader.run()
         callback = self.mock_mqtt_client.on_disconnect
         callback(self.mock_mqtt_client, None, 1)
         assert "MQTT client has disconnected: 1" == str(caplog.records[0].message)
+
+    def test_unhandled_exception(self):
+        def process_message(json_str):
+            raise RuntimeError("didn't expect that!")
+
+        self.mqtt_reader.add_message_callback(process_message)
```

Comment:
oh, I guess the callbacks were added to make the testing easier, but I wonder if we can get around that by just using common DI instead. Injecting a mock that enables the same expected behaviour

Reply:
I think the main benefit of the `add_message_callback` is to keep `MQTTReader` generic and allow any module to use it and perform any action on receiving a message. If we don't want to keep it that way, we can definitely work around not having this callback method in tests.